devdoc/roxie.md

```diff
@@ -189,4 +189,45 @@ Having first compiled a suitable bit of ECL into a.out. I have found a snippet l
     END;
     
     d := dataset([{rtl.sleep(5000)}], {unsigned a});
-    allnodes(d)+d;
\ No newline at end of file
+    allnodes(d)+d;
+
+## Cache prewarm
+
+Roxie (optionally) maintains a list of the most recently accessed file pages (in a circular buffer), and flushes this information
+periodically to text files that will persist from one run of Roxie to the next. On startup, these files are processed and the
+relevant pages preloaded into the linux page cache to ensure that the "hot" pages are already available and maximum performance is
+available immediately once the Roxie is brought online, rather than requiring a replay of a "typical" query set to heat the cache
+as used to be done. In particular this should allow a node to be "warm" before being added to the cluster when autoscaling.
+
+There are some questions outstanding about how this operates that may require empirical testing to answer. Firstly, how does this
+interact with volumes mounted via k8s pvc's, and in particular with cloud billing systems that charge per read. Will the reads
+that are done to warm the cache be done in large chunks, or will they happen one linux page at a time? The code at the Roxie level
+operates by memory-mapping the file then touching a byte within each linux page that we want to be "warm", but does the linux paging
+subsystem fetch larger blocks? Do huge pages play a part here?
+
+Secondly, the prewarm is actually done by a child process (ccdcache), but the parent process is blocked while it happens. It would
+probably make sense to at allow at least some of the other startup operations of the parent process to proceed in parallel. I don't
+recall why it was done via a child process - I suspect that it was coded to allow the prewarm to be done manually (e.g. in the bare-
```

Comment:
My memory is that it is something to do with catching segment faults that can occur, and making sure they don't terminate the roxie process.

Reply:
Ah, that would make sense. I will update the doc accordingly.